chia/util/db_wrapper.py

```diff
@@ -29,3 +33,96 @@ async def rollback_transaction(self):
+    _current_writer: Optional[asyncio.Task]
+
+    async def add_connection(self, c: aiosqlite.Connection) -> None:
+        # this guarnatees that reader connections can only be used for reading
+        await c.execute("pragma query_only")
+        self._read_connections.put_nowait(c)
+        self._num_read_connections += 1
+
+    def __init__(self, connection: aiosqlite.Connection, db_version: int = 1) -> None:
+        self._read_connections = asyncio.Queue()
+        self._write_connection = connection
+        self._lock = asyncio.Lock()
+        self.db_version = db_version
+        self._num_read_connections = 0
+        self._in_use = {}
+        self._current_writer = None
+
+    async def close(self) -> None:
+        while self._num_read_connections > 0:
+            await (await self._read_connections.get()).close()
+            self._num_read_connections -= 1
+        await self._write_connection.close()
+
+    @contextlib.asynccontextmanager
+    async def write_db(self) -> AsyncIterator[aiosqlite.Connection]:
+        task = asyncio.current_task()
+        assert task is not None
+        if self._current_writer == task:
+            # we allow nesting writers within the same task
+            yield self._write_connection
+            return
+
+        async with self._lock:
+
+            global savepoint_name
+            name = f"s{savepoint_name}"
+            savepoint_name += 1
+
+            await self._write_connection.execute(f"SAVEPOINT {name}")
+            try:
+                self._current_writer = task
+                yield self._write_connection
+            except BaseException:
+                await self._write_connection.execute(f"ROLLBACK TO {name}")
+                # rollback to a savepoint doesn't cancel the transaction, it
+                # just restarts it at its initial state. We want to cancel it
+                # here, so we release it as well
+                await self._write_connection.execute(f"RELEASE {name}")
+                raise
+            finally:
+                self._current_writer = None
+            await self._write_connection.execute(f"RELEASE {name}")
+
+    @contextlib.asynccontextmanager
+    async def read_db(self) -> AsyncIterator[aiosqlite.Connection]:
+        # there should have been read connections added
+        assert self._num_read_connections > 0
+
+        # we can have multiple concurrent readers, just pick a connection from
+        # the pool of readers. If they're all busy, we'll wait for one to free
+        # up.
+        task = asyncio.current_task()
+        assert task is not None
+
+        # if this task currently holds the write lock, use the same connection,
+        # so it can read back updates it has made to its transaction, even
+        # though it hasn't been comitted yet
+        if self._current_writer == task:
+            # we allow nesting writers within the same task
+            yield self._write_connection
```

Comment:
This bypasses the read-only nature of the reader connection you would otherwise get.  Can that pragma be applied 'temporarily' while this is yielded here?

Reply:
you're right. I'm a bit worried about inserting calls to `pragma query_only=1` and `pragma query_only=0` frequently though.

Reply:
maybe we should have a `test_db` or `strict_db` setting that would do this, as well as setting things like `pragma reverse_unordered_selects`

Reply:
Aside from the pragma topic, I guess this indicates that we should force the practice of closing _all_ cursors.  Easiest way is to just require use of a context manager consistently for _all_ `.execute()` calls.

1. We could yield a custom 'connection' object that only offers context manager execute mode
2. That custom 'connection' could complain about `pragma` attempts
3. There could be direct `async with db_wrapper.write_execute()` and `read` methods for when you only want to run one

I'll put the pragma topic on my list of followup considerations.